test/gtest/graphapi_engine.cpp

```diff
@@ -0,0 +1,134 @@
+    {
+    }
+    size_t getWorkspaceSize() const override { return 0; }
+};
+
+} // namespace
+
+TEST(GraphApi, EngineBuilder)
+{
+    OpGraph opGraph;
+    auto executor = std::make_shared<MockPatternExecutor>();
+
+    EXPECT_NO_THROW({
+        EngineBuilder().setGraph(&opGraph).setGlobalIndex(0).setExecutor(executor).build();
+    }) << "EngineBuilder failed on valid attributes";
+
+    EXPECT_ANY_THROW({ EngineBuilder().setGlobalIndex(0).setExecutor(executor).build(); })
+        << "EngineBuilder failed on missing setGraph() call";
+
+    EXPECT_ANY_THROW({ EngineBuilder().setGraph(&opGraph).setExecutor(executor).build(); })
+        << "EngineBuilder failed on missing setGlobalIndex() call";
+
+    EXPECT_ANY_THROW({ EngineBuilder().setGraph(&opGraph).setGlobalIndex(0).build(); })
+        << "EngineBuilder failed on missing setExecutor() call";
+}
+
+namespace {
+
+using miopen::graphapi::GTestDescriptorAttribute;
+using miopen::graphapi::GTestDescriptorSingleValueAttribute;
+using miopen::graphapi::GTestGraphApiExecute;
+
+class MockOpGraphDescriptor : public miopen::graphapi::BackendOperationGraphDescriptor
+{
+public:
+    MockOpGraphDescriptor() { mFinalized = true; }
+};
+
+} // namespace
+
+TEST(GraphApi, Engine)
+{
+    MockOpGraphDescriptor opGraphDescriptor;
+
+    GTestDescriptorSingleValueAttribute<miopenBackendDescriptor_t, char> attrOpGraph(
+        true,
+        "MIOPEN_ATTR_ENGINE_OPERATION_GRAPH",
+        MIOPEN_ATTR_ENGINE_OPERATION_GRAPH,
+        MIOPEN_TYPE_BACKEND_DESCRIPTOR,
+        MIOPEN_TYPE_CHAR,
+        2,
+        &opGraphDescriptor);
+
+    GTestDescriptorSingleValueAttribute<int64_t, char> attrGlobalIndex(
+        true,
+        "MIOPEN_ATTR_ENGINE_GLOBAL_INDEX",
+        MIOPEN_ATTR_ENGINE_GLOBAL_INDEX,
+        MIOPEN_TYPE_INT64,
+        MIOPEN_TYPE_CHAR,
+        2,
+        0);
+
+    GTestDescriptorSingleValueAttribute<int32_t, char> attrSmCount(
+        true,
+        "MIOPEN_ATTR_ENGINE_SM_COUNT_TARGET",
+        MIOPEN_ATTR_ENGINE_SM_COUNT_TARGET,
+        MIOPEN_TYPE_INT32,
+        MIOPEN_TYPE_CHAR,
+        2,
+        1);
+
+    GTestGraphApiExecute<GTestDescriptorAttribute*> execute;
+
+    /* BackendOperationGraphDescriptor and the whole API design lacks dependency inversion
+     * to be able to achieve a full coverage in tests for BackendOperationGraphDescriptor
+     * without touching other parts of the code. We cannot mock OpGraph and make it return
+     * a non-empty mock list of Engines, so here we test only attrsValid=false case.
+     *
+     * We have other tests (for example MHA) that cover attrsValid=true case but they
+     * cover the whole API code.
+     */
+
+    execute.descriptor.attrsValid = false;
+    execute.descriptor.textName   = "MIOPEN_BACKEND_ENGINE_DESCRIPTOR";
+    execute.descriptor.type       = MIOPEN_BACKEND_ENGINE_DESCRIPTOR;
+    execute.descriptor.attributes = {&attrOpGraph, &attrGlobalIndex, &attrSmCount};
+
+    execute();
```

Comment:
Similar to my comments on the other PR, we should add tests for replacing `attrOpGraph` and `attrGlobalIndex`.

Reply:
Yes, you are right.

`attrGlobalIndex` becomes invalid here because mock `OpGraph` returns an empty array of `Engines`. To overcome it we also should somehow mock `patterns` variable in `findEngines()` function.

`BackendOperationGraphDescriptor` class depends on concrete class `OpGraph` which depends on concrete class `MHA_Fwd_F8_Pattern` (through the call of `OpGraph::initEngines()` and `findEngines()`)

The dependency on `MHA_Fwd_F8_Pattern` should be replaced by a dependency on abstract `GraphPatternMatcher` by passing the vector of GraphPatternMatchers through this sequence above of dependent classes. There should be a point where we create a concrete class instances, seems like it is `miopenBackendCreateDescriptor()`.

This refactoring is not the point of this PR and should be planned separately.

Also we can write tests which cover the whole process instead of this particular step. We have MHA test that goes that way.

Reply:
Yes, currently hard to mock patterns and executors. Let me think about it as I'm not sure if it's worth allowing such a thing. We might have a baseline graph executor that executes every valid graph. This may simplify testing here.